src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -796,7 +811,18 @@ module IncrementalBuilderHelpers =
                 mkSimpleAssemblyRef assemblyName, ProjectAssemblyDataResult.Unavailable true, None
 
         let finalBoundModel = Array.last computedBoundModels
-        let diagnostics = diagnosticsLogger.GetDiagnostics() :: finalInfo.tcDiagnosticsRev
+
+        // Collect diagnostics. This will type check in parallel any implementation files skipped so far.
+        let! partialDiagnostics =
+            computedBoundModels
+            |> Seq.map (fun m -> m.Diagnostics.GetOrComputeValue())
+            |> NodeCode.Parallel
+        let diagnostics = [
```

Comment:
Just to double-check, we won't get any duplicate diagnostics using this?

Reply:
Shouldn't be the case because the only things accessible by `Diagnostics` prop are `newErrors` from typechecking single file. From here:
https://github.com/dotnet/fsharp/blob/fa8b61bf2c7f1d88a0cad1e49f65cae98aa7c9d0/src/Compiler/Service/IncrementalBuild.fs#L283-L301

Reply:
Basically instead of incrementally accumulating:
`newErrors :: prevTcInfo.tcDiagnosticsRev`
we do a single pass in `FinalizeTypeCheckTask`.